sqld/src/http/mod.rs

```diff
@@ -193,48 +195,61 @@ async fn handle_fallback() -> impl IntoResponse {
 
 /// Router wide state that each request has access too via
 /// axum's `State` extractor.
-pub(crate) struct AppState<D> {
+pub(crate) struct AppState<F: NamespaceFactory> {
     auth: Arc<Auth>,
-    db_factory: Arc<dyn DbFactory<Db = D>>,
+    namespaces: Arc<Namespaces<F>>,
     upgrade_tx: mpsc::Sender<hrana::ws::Upgrade>,
-    hrana_http_srv: Arc<hrana::http::Server<D>>,
+    hrana_http_srv: Arc<hrana::http::Server<F::Database>>,
     enable_console: bool,
     stats: Stats,
+    allow_default_namespace: bool,
 }
 
-impl<D> Clone for AppState<D> {
+impl<F: NamespaceFactory> Clone for AppState<F> {
     fn clone(&self) -> Self {
         Self {
             auth: self.auth.clone(),
-            db_factory: self.db_factory.clone(),
+            namespaces: self.namespaces.clone(),
             upgrade_tx: self.upgrade_tx.clone(),
             hrana_http_srv: self.hrana_http_srv.clone(),
             enable_console: self.enable_console,
             stats: self.stats.clone(),
+            allow_default_namespace: self.allow_default_namespace,
         }
     }
 }
 
 // TODO: refactor
 #[allow(clippy::too_many_arguments)]
-pub async fn run_http<D: Database>(
+pub async fn run_http<F, S>(
     addr: SocketAddr,
     auth: Arc<Auth>,
-    db_factory: Arc<dyn DbFactory<Db = D>>,
+    namespaces: Arc<Namespaces<F>>,
     upgrade_tx: mpsc::Sender<hrana::ws::Upgrade>,
-    hrana_http_srv: Arc<hrana::http::Server<D>>,
+    hrana_http_srv: Arc<hrana::http::Server<F::Database>>,
     enable_console: bool,
     idle_shutdown_layer: Option<IdleShutdownLayer>,
     stats: Stats,
-    logger: Option<Arc<ReplicationLogger>>,
-) -> anyhow::Result<()> {
+    replication_service: Option<S>,
```

Comment:
I couldn't figure out why this was changed to be generic? Could you expand on that, I think adding generics if we don't need it can make this API much harder to read.

Reply:
2 step plan:
- don't leak namespace-specific information to HTTP (`Namespace<PrimaryNamespaceFactory>`)
- rn, only the primary can serve replication requests, but, it's easy enough to make a proxy rpc service so that replication request can be served from any node, which is what I want to do in a followup PR.

Reply:
Would it be better to just enum that service rather than make things more generic?